Add workspace UI

Gkrumbach07 opened the pull request:
### Related Issues/PRs

#1464
#5844
Design document: https://docs.google.com/document/d/1IbFfceCmWV3knJfc0ninn58EXLVbHUh1meV_pknOM40/edit

### What changes are proposed in this pull request?

This PR adds the frontend UI components for multi-workspace support in MLflow.

**⚠️ Important: Backward Compatibility**
The new workspace UI behavior is **only activated** when the server's `/ajax-api/2.0/mlflow/server-features` endpoint returns `{ "workspacesEnabled": true }`. When workspaces are not enabled (default behavior), the UI operates exactly as before with no visible changes.

**New Components:**
- `WorkspaceSelector` - Dropdown component with search and tooltips for selecting workspaces
- `WorkspacePermissionError` - Error page displayed when accessing an unauthorized workspace
- `useWorkspaces` hook - Custom hook for fetching available workspaces from the API
- `ServerFeaturesContext` - Context provider for detecting server feature flags via `/ajax-api/2.0/mlflow/server-features`
- `WorkspaceUtils` - Utilities for workspace state management and URL handling
- `WorkspaceRouteUtils` - Utilities for prefixing routes with workspace paths

**Core Changes:**
- `MlflowRouter` - Added workspace-aware routing with `WorkspaceRouterSync` component (only active when workspaces enabled)
- `app.tsx` - Integrated `ServerFeaturesProvider` for feature detection
- `FetchUtils` - Added `X-MLFLOW-WORKSPACE` header injection for all API requests (only when workspace is set)
- `graphql/client` - Added workspace header support for GraphQL requests

**Behavior when workspaces are enabled:**
- Frontend URLs include workspace: `/#/workspaces/team-a/experiments`
- Backend API calls pass workspace via `X-MLFLOW-WORKSPACE` header (not URL path)
- Workspace selector appears in header
- Automatic redirect to default workspace when none is specified
- Workspace state persisted to localStorage

**Behavior when workspaces are disabled (default):**
- No workspace selector shown
- No workspace prefix in URLs
- No `X-MLFLOW-WORKSPACE` header sent
- UI operates exactly as before

<img width="1196" height="605" alt="image" src="https://github.com/user-attachments/assets/865e9dd3-8d4a-4dc0-bbc5-ffac99bcaf50" />
<img width="464" height="236" alt="image" src="https://github.com/user-attachments/assets/58b3c53a-fdec-4f16-b776-1ae01ebc7e4b" />


### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests

New tests added for:
- `ServerFeaturesContext.test.tsx` - Tests feature detection and workspace enabled/disabled states
- `WorkspaceRouteUtils.test.ts` - Tests route prefixing logic
- `ArtifactUtils.test.ts` - Tests workspace header injection
- `graphql/client.test.ts` - Tests GraphQL workspace header support
- Updated existing tests to mock workspace context

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.

Documentation will be added in a follow-up PR once the full workspace feature is complete.

### Release Notes

#### Is this a user-facing change?

- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Added multi-workspace UI support allowing users to switch between different workspaces via a dropdown selector in the MLflow header. This feature is only enabled when the MLflow server is configured with workspace support (determined by the `/ajax-api/2.0/mlflow/server-features` endpoint). When workspaces are not enabled, the UI behaves exactly as before.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [x] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

Gkrumbach07 commented:
Here is the new landing page
<img width="1512" height="863" alt="Screenshot 2026-01-19 at 10 10 13 PM" src="https://github.com/user-attachments/assets/be447181-0055-4d36-93e2-2e6381d4bb0c" />
<img width="819" height="605" alt="Screenshot 2026-01-19 at 10 21 20 PM" src="https://github.com/user-attachments/assets/ae8bc5d3-cb1f-426b-bd34-b0201a8a3eac" />

note:
- i removed the notion of navigating to a default/last used workspace automatically otherwise this workspace page would never get seen

Gkrumbach07 commented:
- Fixed create experiment not showing error
- hide artifact location in experiment create modal when in workspace mode
- artifact root added to workspace create modal and table
- able to edit description and artifact root via workspace page table edit buttons inline
- route now uses root instead of /workspaces for workspace landing page
- can navigate back to workspace landing page via side nav

<img width="1811" height="959" alt="Screenshot 2026-01-20 at 4 28 17 PM" src="https://github.com/user-attachments/assets/3fa0eae2-076f-4966-8021-636f1239e046" />
<img width="1811" height="959" alt="Screenshot 2026-01-20 at 4 28 01 PM" src="https://github.com/user-attachments/assets/9cdfdca7-4afc-4517-81ef-d36a9abdf7d7" />
<img width="1811" height="959" alt="Screenshot 2026-01-20 at 4 27 55 PM" src="https://github.com/user-attachments/assets/638776c3-94ea-4999-9d83-aea22b25850f" />
<img width="1811" height="959" alt="Screenshot 2026-01-20 at 4 27 42 PM" src="https://github.com/user-attachments/assets/8896a833-9ee8-4847-adf8-7d35034d5401" />
<img width="1811" height="959" alt="Screenshot 2026-01-20 at 4 29 23 PM" src="https://github.com/user-attachments/assets/1f68b878-723e-4461-af48-97c0ef042dee" />

mprahl commented on mlflow/server/js/src/common/utils/WorkspaceUtils.ts:
```diff
@@ -0,0 +1,228 @@
+import { getWorkspacesEnabledSync } from './ServerFeaturesContext';
```
Let's simplify andreduce duplication in this file. Here is a suggestion from Claude/GPT:
There are a few simplification opportunities in `WorkspaceUtils.ts`:

- `isWorkspacesFeatureEnabled()` is a passthrough to `getWorkspacesEnabledSync()`; can be inlined or removed.
- `availableWorkspaces` / `getAvailableWorkspaces()` are only referenced in tests (getter unused in prod); consider removing or clarifying intended usage.
- `prefixRouteWithWorkspace` and `prefixPathnameWithWorkspace` duplicate logic; a small helper for building `/workspaces/${encodeURIComponent(...)}` paths would reduce repetition.
- `sanitizePath()` is tiny and only used twice; could be inlined if you want fewer helpers.
- Repeated `encodeURIComponent(workspace)` can be centralized.

mprahl commented:
In the `Configure experiment and tracking URI` section in `mlflow/server/js/src/home/components/LogTracesDrawer.tsx`, could you conditionally add the `mlflow.set_workspace(<workspace>)` if there is a workspace selected?

Same for `mlflow/server/js/src/experiment-tracking/components/evaluation-artifacts-compare/CreateNotebookRunModal.tsx`

mprahl commented:
A slightly annoying thing is I can't hit the "Enter" key to submit the forms in the new workspace modals (create and edit).

Gkrumbach07 commented on mlflow/server/js/src/common/components/WorkspacePermissionError.tsx:
```diff
@@ -0,0 +1,25 @@
+import { ErrorView } from './ErrorView';
+import { WorkspaceSelector } from './WorkspaceSelector';
+
+export const WorkspacePermissionError = ({ workspaceName }: { workspaceName: string }) => {
```
its old from past handling. i will remove it

Gkrumbach07 commented on mlflow/server/js/src/common/utils/FetchUtils.ts:
```diff
@@ -484,15 +510,29 @@ export const fetchAPI = async (url: string, method: 'POST' | 'GET' | 'PATCH' | '
 /**
  * Wrapper around fetch that throws on non-OK responses
  * Returns the Response object for further processing (.json(), .text(), etc.)
+ * Automatically includes default headers (cookies, auth, and workspace).
```
done

mprahl commented:
/lgtm

hubertzub-db commented on mlflow/server/js/src/common/components/WorkspaceSelector.tsx:
```diff
@@ -0,0 +1,178 @@
+    }
+  };
+
+  const options = useMemo(() => {
```
nit: it's possible to use lodash's `uniqBy` for clarity

Gkrumbach07 commented on mlflow/server/js/src/shared/web-shared/genai-traces-table/utils/FetchUtils.ts:
```diff
@@ -1,75 +1,20 @@
-import { matchPredefinedError } from '../../errors';
+import {
```
thats fine, i will fix it

Gkrumbach07 committed:
feat(ui): Add workspace landing page, multi-workspace support, and query param routing

Signed-off-by: Gage Krumbach <gkrumbach@gmail.com>
Co-Authored-By: Matt Prahl <mprahl@users.noreply.github.com>

hubertzub-db reviewed (commented):
Took another pass, this looks good! Thanks a lot for redoing everything to use query params instead of a prefix!
- Left a few minor questions
- A major one: what's the exact issue/error that made you add `<TooltipProvider>` to many tests in the app? In general, `<TooltipProvider>` should be automatically added by `<DesignSystemProvider>` - if it doesn't, it indicates some deeper problem in design system we'd like to investigate
- CC @daniellok-db would you like to take a quick look at this PR as well?

Gkrumbach07 committed:
fix(ui): Address workspace PR review feedback

- Remove redundant TooltipProvider from all test files (DesignSystemProvider
  already includes RadixTooltipProvider internally)
- Refactor workspace switching from pub/sub to hard reload for simplicity
  and to eliminate redirect race conditions
- Restore fetchFn abstraction in FetchUtils.ts for codebase sync
- Remove stale workspace routing comment
- Fix upstream bug in useMlflowTraces.test.tsx (fetchFn -> fetchAPI)

Signed-off-by: Gage Krumbach <gkrumbach@gmail.com>

Gkrumbach07 commented on mlflow/server/js/src/workspaces/utils/WorkspaceUtils.ts:
```diff
@@ -0,0 +1,243 @@
+
+export const setActiveWorkspace = (workspace: string | null) => {
+  activeWorkspace = workspace;
+  listeners.forEach((listener) => listener(activeWorkspace));
```
i did the refresh pattern

Gkrumbach07 commented:
**Addressed all feedback from the latest review:**

1. **`TooltipProvider` in tests**: Investigated and confirmed `DesignSystemProvider` already includes `RadixTooltipProvider` internally (they resolve to the same `@radix-ui/react-tooltip`). The extra wrappers were unnecessary so i removed them all

2. **`fetchFn` abstraction** — Restored `const fetchFn = fetch` in `FetchUtils.ts` to keep in sync with the internal codebase.

3. **Stale comment** in `useGetExperimentPageActiveTabByRoute.tsx` -> Removed.

4. **Pub/sub → hard reload** -> Agreed this is the better trade off. Removed the `listeners`/`subscribeToWorkspaceChanges` pattern entirely. Workspace switching now does `window.location.hash` + `window.location.reload()`, which naturally handles cache invalidation

hubertzub-db reviewed (approved):
@Gkrumbach07 thanks for all the changes and fixes!
Stamping as it looks good in the current form, I'll defer to @daniellok-db for further review

daniellok-db reviewed (approved):
Overall LGTM but there seems to be some bug with links to the chat sessions page, e.g.

https://github.com/mlflow/mlflow/blob/c551c5232179127d9dcc5f9847bbaa34a9cb5bc4/mlflow/server/js/src/shared/web-shared/model-trace-explorer/ModelTraceHeaderSessionIdTag.tsx#L76

My suspicion is that this is due to the packages in `src/shared/web-shared` having their own individual copies of `RoutingUtils`. Unfortunately this is a quirk of our setup where the UI has some upstream dependencies (we cannot import from `@mlflow/mlflow` from within that folder as it will cause conflicts). See:

https://github.com/mlflow/mlflow/blob/c551c5232179127d9dcc5f9847bbaa34a9cb5bc4/mlflow/server/js/src/shared/web-shared/genai-traces-table/utils/RoutingUtils.tsx#L4

https://github.com/mlflow/mlflow/blob/c551c5232179127d9dcc5f9847bbaa34a9cb5bc4/mlflow/server/js/src/shared/web-shared/model-trace-explorer/RoutingUtils.tsx#L4

Could you copy paste the changes you made over to those files too?

Doesn't have to block the merge, I'll leave it up to @B-Step62 but we should definitely follow up at least in a future PR
